moler/observable_connection.py

```diff
@@ -0,0 +1,179 @@
+
+import weakref
+import logging
+import six
+from threading import Lock
+from moler.connection import Connection
+from moler.connection import identity_transformation
+from moler.config.loggers import RAW_DATA, TRACE
+from moler.helpers import instance_id
+
+
+class ObservableConnection(Connection):
+    """
+    Allows objects to subscribe for notification about connection's data-received.
+    Subscription is made by registering function to be called with this data (may be object's method).
+    Function should have signature like:
+
+    def observer(data):
+        # handle that data
+    """
+
+    def __init__(self, how2send=None, encoder=identity_transformation, decoder=identity_transformation,
+                 name=None, newline='\n', logger_name=""):
+        """
+        Create Connection via registering external-IO
+
+        :param how2send: any callable performing outgoing IO
+        :param encoder: callable converting data to bytes
+        :param decoder: callable restoring data from bytes
+        :param name: name assigned to connection
+        :param logger_name: take that logger from logging
+
+        Logger is retrieved by logging.getLogger(logger_name)
+        If logger_name == "" - take logger "moler.connection.<name>"
+        If logger_name is None - don't use logging
+        """
+        super(ObservableConnection, self).__init__(how2send, encoder, decoder, name=name, newline=newline,
+                                                   logger_name=logger_name)
+        self._observers = dict()
+        self._connection_closed_handlers = dict()
+        self._observers_lock = Lock()
+
+    def data_received(self, data):
+        """
+        Incoming-IO API:
+        external-IO should call this method when data is received
+        """
+        if not self.is_open():
+            return
+        extra = {'transfer_direction': '<', 'encoder': lambda data: data.encode(encoding='utf-8', errors="replace")}
+        self._log_data(msg=data, level=RAW_DATA,
+                       extra=extra)
+
+        decoded_data = self.decode(data)
+        self._log_data(msg=decoded_data, level=logging.INFO,
+                       extra=extra)
+
+        self.notify_observers(decoded_data)
+
+    def subscribe(self, observer, connection_closed_handler):
+        """
+        Subscribe for 'data-received notification'
+
+        :param observer: function to be called to notify when data received.
+        :param connection_closed_handler: callable to be called when connection is closed.
+        """
+        with self._observers_lock:
+            self._log(level=TRACE, msg="subscribe({})".format(observer))
+            observer_key, value = self._get_observer_key_value(observer)
+
+            if observer_key not in self._observers:
+                self._observers[observer_key] = value
+                self._connection_closed_handlers[observer_key] = connection_closed_handler
```

Comment:
instead of registering "close-callback" subscriptions inside Moler-connection:

- let moler connection subscribe that callback inside io-connection (we already have that functionality)
- then we don't need close here because io-connection close() will trigger callbacks